Add support for realm-level defaults of personal settings for new users.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Commits are -
- Do not translate messages sent to "zulip internal" realm (as discussed on chat.zulip.org)
- Remove unnecessary fields from `realm/update_dict` event (#18697).
- Add a `UserBaseSettings` class
- Add `RealmUserDefault` table.
- Create the `RealmUserDefault` table when creating realms.
- Migration to create `RealmUserDefault` for existing realms.
- Move the realm fields `default_language` and `default_twenty_four_hour_time` to `RealmUserDefault` table.

Related issue - #11570
Related discussion - https://chat.zulip.org/#narrow/stream/9-issues/topic/add.20realm.20default.20value.20for.20user.20settings.20.2311570/near/1198714

**Testing plan:** <!-- How have you tested? --> Updated the existing tests for last commit. 


 <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

sahil839 commented:
Few points to note/discuss -
- Right now, have added only display settings (considering settings which are updated using `update_display_settings_backend` as display settings) except timezone as discussed on czo.
- The last commit adds a lot of queries mostly for `default_langauge` uses, can we eliminate that by adding `realm_user_default` as a foreign key to `UserProfile` similar to `realm`.
- I have currently kept the event type as `realm_default` for these settings.
- Do we need a separate endpoint for changing these settings like `PATCH /realm_default` or something else. Right now have kept the endpoints for `default_langauge` and `twenty_four_hour_time` same, ie. `PATCH /realm`.

sahil839 commented on zerver/lib/export.py:
```diff
@@ -142,6 +142,7 @@
     "zerver_realmemoji",
     "zerver_realmfilter",
     "zerver_realmplayground",
+    "zerver_realmuserdefault",
```
I am not sure we would want to export this table, but just included it for now.

timabbott replied:
Yes, we definitely do.  We should also be sure to import it in the `import_realm.py` logic.

sahil839 commented on zerver/tests/test_signup.py:
```diff
@@ -1065,7 +1067,7 @@ def try_invite() -> HttpResponse:
         #       use an approximate equality check.
         actual_count = len(queries)
-        expected_count = 251
+        expected_count = 313
```
Here we send 30 invites, so according to 2 query per invite (one for invitation email and other for invitation reminder email), there should be increase of 60 queries, not sure of the reason of 2 additional queries.

timabbott replied:
This is a pretty big increase!  I think we can fix this with caching, or an optional parameter to the function to send confirmation emails (to avoid fetching the same parameter N times in a loop, which is what we're doing here).

I'm not sure why this number is 250+ in the first place; but I guess there's a TODO comment.  It would be a good use of time to debug that in a separate PR; for this one, I think it's probably at least worth avoiding another another 60.

sahil839 replied:
```diff
diff --git a/zerver/lib/actions.py b/zerver/lib/actions.py
index c4e66e66d2..b01177e373 100644
--- a/zerver/lib/actions.py
+++ b/zerver/lib/actions.py
@@ -6624,7 +6624,7 @@ def filter_presence_idle_user_ids(user_ids: Set[int]) -> List[int]:
     return sorted(idle_user_ids)
 
 
-def do_send_confirmation_email(invitee: PreregistrationUser, referrer: UserProfile) -> str:
+def do_send_confirmation_email(invitee: PreregistrationUser, referrer: UserProfile, language: str) -> str:
     """
     Send the confirmation/welcome e-mail to an invited user.
     """
@@ -6635,12 +6635,11 @@ def do_send_confirmation_email(invitee: PreregistrationUser, referrer: UserProfi
         "activate_url": activation_url,
         "referrer_realm_name": referrer.realm.name,
     }
-    realm_user_default = RealmUserDefault.objects.get(realm=referrer.realm)
     send_email(
         "zerver/emails/invitation",
         to_emails=[invitee.email],
         from_address=FromAddress.tokenized_no_reply_address(),
-        language=realm_user_default.default_language,
+        language=language,
         context=context,
         realm=referrer.realm,
     )
@@ -6791,6 +6790,7 @@ def do_invite_users(
         increment=len(validated_emails),
     )
 
+    realm_user_default = RealmUserDefault.objects.get(realm=user_profile.realm)
     # Now that we are past all the possible errors, we actually create
     # the PreregistrationUser objects and trigger the email invitations.
     for email in validated_emails:
@@ -6802,7 +6802,7 @@ def do_invite_users(
         stream_ids = [stream.id for stream in streams]
         prereg_user.streams.set(stream_ids)
 
-        event = {"prereg_id": prereg_user.id, "referrer_id": user_profile.id}
+        event = {"prereg_id": prereg_user.id, "referrer_id": user_profile.id, "email_language": realm_user_default.default_language}
         queue_json_publish("invites", event)
 
     if skipped:
@@ -6920,11 +6920,13 @@ def do_resend_user_invite_email(prereg_user: PreregistrationUser) -> int:
diff --git a/zerver/lib/actions.py b/zerver/lib/actions.py
index c4e66e66d2..b01177e373 100644
--- a/zerver/lib/actions.py
+++ b/zerver/lib/actions.py
@@ -6624,7 +6624,7 @@ def filter_presence_idle_user_ids(user_ids: Set[int]) -> List[int]:
     return sorted(idle_user_ids)
 
 
-def do_send_confirmation_email(invitee: PreregistrationUser, referrer: UserProfile) -> str:
+def do_send_confirmation_email(invitee: PreregistrationUser, referrer: UserProfile, language: str) -> str:
     """
     Send the confirmation/welcome e-mail to an invited user.
     """
@@ -6635,12 +6635,11 @@ def do_send_confirmation_email(invitee: PreregistrationUser, referrer: UserProfi
         "activate_url": activation_url,
         "referrer_realm_name": referrer.realm.name,
     }
-    realm_user_default = RealmUserDefault.objects.get(realm=referrer.realm)
     send_email(
         "zerver/emails/invitation",
         to_emails=[invitee.email],
         from_address=FromAddress.tokenized_no_reply_address(),
-        language=realm_user_default.default_language,
+        language=language,
         context=context,
         realm=referrer.realm,
     )
@@ -6791,6 +6790,7 @@ def do_invite_users(
         increment=len(validated_emails),
     )
 
+    realm_user_default = RealmUserDefault.objects.get(realm=user_profile.realm)
     # Now that we are past all the possible errors, we actually create
     # the PreregistrationUser objects and trigger the email invitations.
     for email in validated_emails:
@@ -6802,7 +6802,7 @@ def do_invite_users(
         stream_ids = [stream.id for stream in streams]
         prereg_user.streams.set(stream_ids)
 
-        event = {"prereg_id": prereg_user.id, "referrer_id": user_profile.id}
+        event = {"prereg_id": prereg_user.id, "referrer_id": user_profile.id, "email_language": realm_user_default.default_language}
         queue_json_publish("invites", event)
 
     if skipped:
@@ -6920,11 +6920,13 @@ def do_resend_user_invite_email(prereg_user: PreregistrationUser) -> int:
     )
 
     clear_scheduled_invitation_emails(prereg_user.email)
+    realm_user_default = RealmUserDefault.objects.get(realm=prereg_user.realm)
     # We don't store the custom email body, so just set it to None
     event = {
         "prereg_id": prereg_user.id,
         "referrer_id": prereg_user.referred_by.id,
         "email_body": None,
+        "email_language": realm_user_default.default_language,
     }
     queue_json_publish("invites", event)
 
diff --git a/zerver/tests/test_signup.py b/zerver/tests/test_signup.py
index c8cb3fa2f1..a95cd8b2f3 100644
--- a/zerver/tests/test_signup.py
+++ b/zerver/tests/test_signup.py
@@ -1067,7 +1067,7 @@ class InviteUserTest(InviteUserBase):
         #       the large number of queries), so I just
         #       use an approximate equality check.
         actual_count = len(queries)
-        expected_count = 313
+        expected_count = 253
         if abs(actual_count - expected_count) > 1:
             raise AssertionError(
                 f"""
diff --git a/zerver/worker/queue_processors.py b/zerver/worker/queue_processors.py
index d1e3230735..3d9b7ebdd8 100644
--- a/zerver/worker/queue_processors.py
+++ b/zerver/worker/queue_processors.py
@@ -425,7 +425,7 @@ class ConfirmationEmailWorker(QueueProcessingWorker):
         logger.info(
             "Sending invitation for realm %s to %s", referrer.realm.string_id, invitee.email
         )
-        activate_url = do_send_confirmation_email(invitee, referrer)
+        activate_url = do_send_confirmation_email(invitee, referrer, data["email_language"])
 
         # queue invitation reminder
         if settings.INVITATION_LINK_VALIDITY_DAYS >= 4:
@@ -436,13 +436,12 @@ class ConfirmationEmailWorker(QueueProcessingWorker):
                 referrer_email=referrer.delivery_email,
                 referrer_realm_name=referrer.realm.name,
             )
-            realm_user_default = RealmUserDefault.objects.get(realm=referrer.realm)
             send_future_email(
                 "zerver/emails/invitation_reminder",
                 referrer.realm,
                 to_emails=[invitee.email],
                 from_address=FromAddress.tokenized_no_reply_placeholder,
-                language=realm_user_default.default_language,
+                language=data["email_language"],
                 context=context,
                 delay=datetime.timedelta(days=settings.INVITATION_LINK_VALIDITY_DAYS - 2),
             )
```

Applying this diff on the current PR state, leads to increase of 2 queries (i.e. 253).

timabbott replied:
Yeah, that looks great.  Maybe we can do that as a prep commit in a separate PR?  We'll need some code to handle old events already in the queue.

sahil839 replied:
Opened #18950.

sahil839 commented:
Have not manually tested the last commit (moving `default_language` and `twent_four_hour_time` to `RealmUserDefault`) yet, other commits can be reviewed though.

sahil839 commented:
I have tested the last commit, works fine. I am not sure about the rename of `default_twenty_four_hour_time`. As we are using a single base class, the same field name will be used for both the user-level setting and realm-level default. I feel `default_twenty_four_hour_time` will not be suitable for user-level setting, as it is not a default value, it is just a setting controlled by user. And we will anyways always access `twenty_four_hour_time` for realm-level default using the appropriate object name like `realm_user_default`, so it will be somewhat clear that it is a realm-level default. I have not renamed all the instances of this field in frontend and in the page_params data sent from server, we can do so if we are happy with the change and have changed the parameter passed to `PATCH /realm` just to avoid extra if conditions for appropriately renaming the name of property while actually updating the DB value and in tests.

Have also left 2-3 comments before this. @timabbott this can be reviewed.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -3305,15 +3305,6 @@ paths:
-                                      type: boolean
-                                      description: |
-                                        Whether new members of this organization will see times displayed in
-                                        24-hour time (true) or 12-hour time (false).
```
One strategic question we should potentially talk about is whether we want to migrate everything that is presently in the `update` format to use `update_dict`, since that's the more generic format; but that'll require some coordination with clients in the "api design" stream.  So I think it's correct to merge this and think about that later.

timabbott commented on zerver/models.py:
```diff
@@ -1147,7 +1147,61 @@ def __str__(self) -> str:
 
 
-class UserProfile(AbstractBaseUser, PermissionsMixin):
+class UserBaseSettings(models.Model):
```
What's the thinking behind not moving "notification settings"?  I suppose we can always move those later and this helps manage complexity, but it'd be good to say that explicitly if that's the intent.

timabbott commented on zerver/models.py:
```diff
@@ -1201,6 +1201,11 @@ class Meta:
         abstract = True
 
 
+class RealmUserDefault(UserBaseSettings):
```
Let's add some documentation here, as well.

timabbott commented on zilencer/management/commands/populate_db.py:
```diff
@@ -318,6 +319,7 @@ def handle(self, **options: Any) -> None:
                 event_time=zulip_realm.date_created,
             )
             RealmDomain.objects.create(realm=zulip_realm, domain="zulip.com")
+            RealmUserDefault.objects.create(realm=zulip_realm)
```
This would really benefit from our doing `do_create_realm` here; @hackerkid what's the status of that migration PR?  It's possible you and @sahil839 collaborate to try to get that finished to help simplify this PR.

hackerkid replied:
#18456 is now ready for a review :)

timabbott commented on zerver/migrations/0332_create_realm_user_default_table.py:
```diff
@@ -0,0 +1,27 @@
+def create_realm_user_default_table(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    Realm = apps.get_model("zerver", "Realm")
+    RealmUserDefault = apps.get_model("zerver", "RealmUserDefault")
+    realms = Realm.objects.exclude(string_id=settings.SYSTEM_BOT_REALM)
```
This deserves a comment on why we're excluding `SYSTEM_BOT_REALM`.

timabbott commented on static/js/server_events_dispatch.js:
```diff
@@ -321,6 +319,17 @@ export function dispatch_normal_event(event) {
+            if (event.property === "twenty_four_hour_time") {
+                property_name = "default_twenty_four_hour_time";
+            }
+            page_params["realm_" + property_name] = event.value;
```
Hmm, what do you think about trying to put these in a `realm_default` object in the top-level of `page_params`, so that this would be just `page.params.realm_default[property_name] = event.value`?

sahil839 replied:
Yes, I though of doing a separate commit for this.

timabbott commented:
I merged the first two commits as the series ending with 5a47e1f2897f7f055a393a6a5a835cd90350bd26, thanks @sahil839!  

The comments above are I think a reasonably complete set of feedback on this PR, so I'm feeling pretty excited about it.  We should do the following:
* Start a conversation on "api design" around the API changes here, specifically adding a new `realm_default` object and its shape, since that's the type of API design detail that is worth getting a lot of eyes on.  E.g. I'm not sure whether we'd prefer `realm_user_defaults` or `realm_default_settings` or something else.  
* Start a second conversation on "api design" about how we want to manage situations where we change the global default for a setting; will the plan be that such migrations will edit the `RealmUserDefault` values for every realm that has the old default?  I think that's reasonable, but wanted to flag it as a design detail to consider, since there's an alternative design that makes every entry in `RealmUserDefault` nullable.
* Do a more thorough audit of which fields we want to allow controlling at the realm level.

One thing that's convenient about this API, though, is that I think mobile/terminal clients likely won't interact with it.

sahil839 commented:
@timabbott Have made the following changes -
- Added the logic for importing the new table, though would need to see for imports from different apps whether they have such settings or not and then modify them accordingly. (I touched the export/import part of the code first time, so would be good to have early feedback on it).
- Added the comments and documents as suggested. 
- Also added `notification settings` in the same commit to avoid another migration as suggested.
- Also added a prep commit for minor refactor.

Also, I have started working on adding `realm_user_defaults` object to be sent to clients and also add the code for remaining settings to be set at time of user creation.

I am also thinking that will migrate the `default_langauge` and `twenty_four_hour_time` to the new endpoint which will be added at the later part because currently the `settings_org` is very abstract in the sense that much of the code for getting data and calling the api endpoint is common for different settings, so I think would be better to move it to that endpoint after we add the UI changes and make separate section in the settings overlay for all the other settings and have a separate js module for them.

sahil839 commented:
> @timabbott Have made the following changes -
> 
> * Added the logic for importing the new table, though would need to see for imports from different apps whether they have such settings or not and then modify them accordingly. (I touched the export/import part of the code first time, so would be good to have early feedback on it).
> * Added the comments and documents as suggested.
> * Also added `notification settings` in the same commit to avoid another migration as suggested.
> * Also added a prep commit for minor refactor.
> 
> Also, I have started working on adding `realm_user_defaults` object to be sent to clients and also add the code for remaining settings to be set at time of user creation.
> 
> I am also thinking that will migrate the `default_langauge` and `twenty_four_hour_time` to the new endpoint which will be added at the later part because currently the `settings_org` is very abstract in the sense that much of the code for getting data and calling the api endpoint is common for different settings, so I think would be better to move it to that endpoint after we add the UI changes and make separate section in the settings overlay for all the other settings and have a separate js module for them.

Apart from the updates quoted above, the last push was to update commit messages and also added a commit for initializing the settings for new users from `RealmUserDefault` table.

sahil839 commented:
@timabbott would be nice to get review on this now, the changes are explained in the previous comment. 
Working on further parts.

timabbott commented:
I merged the first commit, 706ec9714c76d98c2ac230c3a5adf35dfd0c438a, to master.  For the next commit, I think that these 4 settings (as well as the `property_types` and `notification_setting_types` objects feel like they should be properties of `UserBaseSettings`, not `UserProfile`, since they seem like they should be shared between the two data structures.  What's your thinking on not having moved them?

```
    # UI vars                                                                                                             
    enter_sends: Optional[bool] = models.BooleanField(null=True, default=False)                                           
                                                                                                                          
    default_language: str = models.CharField(default="en", max_length=MAX_LANGUAGE_ID_LENGTH)                             
    twenty_four_hour_time: bool = models.BooleanField(default=False)                                                      
                                                                                                                          
    # A timezone name from the `tzdata` database, as found in pytz.all_timezones.                                         
    #                                                                                                                     
    # The longest existing name is 32 characters long, so max_length=40 seems                                             
    # like a safe choice.                                                                                                 
    #                                                                                                                     
    # In Django, the convention is to use an empty string instead of NULL/None                                            
    # for text-based fields. For more information, see                                                                    
    # https://docs.djangoproject.com/en/1.10/ref/models/fields/#django.db.models.Field.null.                              
    timezone: str = models.CharField(max_length=40, default="")                                                           
```

(I realize we might want to migrate actual data to these fields later, but it seems like it'd savea future migration to have these fields be created in the `RealmUserDefault` table, even if the plan is to have them be ignored until we restructure the existing features)

timabbott commented on zerver/lib/create_user.py:
```diff
@@ -111,11 +103,25 @@ def create_user_profile(
+
+        for settings_name in UserProfile.notification_setting_types.keys():
+            settings_value = getattr(realm_user_default, settings_name)
+            setattr(user_profile, settings_name, settings_value)
```
We should perhaps add a block comment here making clear how this related to `copy_user_profile`.  I believe the logic is correct -- we want `copy_user_profile` to take precedence over this new feature.  However, it might make sense to refactor this to call `copy_default_settings(default_settings_source: UserBaseSettings)` with either a `UserProfile` or a `RealmUserDefault` function from the same place, which would make it obvious how this works from the code structure.

timabbott commented on zerver/models.py:
```diff
@@ -1147,7 +1147,93 @@ def __str__(self) -> str:
 
-class UserProfile(AbstractBaseUser, PermissionsMixin):
+class UserBaseSettings(models.Model):
+    # display settings
```
We should add a block comment here explaining the purpose of this model a bit, and also mentioning that whenever we change a value declared here, we may need to do a migration to change that value for all existing RealmUserDefault objects as well, if we want that change in defaults to affect new users in existing organizations?

sahil839 commented:
I think we discussed timezone should not vmbe included here because it would be better if the user's OS or browser timezone is used and not the realm-level default.

timabbott commented:
Yeah, that sounds right -- in line with how we've moved it to be a "custom profile field" in the UI.

sahil839 commented:
@timabbott Done the following changes -
- Added block comments as asked.
- Added a commit to remove `timezone` from `UserProfile.property_types` as we do not want to include it in UserBaseSettings as discussed. Other approach could have been adding the removing `timezone` separately whenver we access `property_types` for `RealmUserDefault` related code, but removing from `property_types` seems cleaner.
- Moved `property_types` and `notification_setting_types` to `UserBaseSettings`.
- Added separate commit for handling initialization of `enter_sends`.

timabbott commented:
I merged the first part of this through 6fc9ae995a4288e44e288a1a68c906456b97c563, after adjusting the docstrings for the two new tables.  This includes the migration to create the new table, so hopefully that'll save some future migration renumbering.

I think for the `property_types` fields, we can still access `UserProfile.property_types` and that feels more readable to me; what are your thoughts on that?  If you feel `UserBaseSettings` access is more readable, maybe start a thread in #backend for visibility on that question?  

We need to take some care in the data import code path; I think as implemented, realms imported via the Slack etc. import tools will violate the invariant of every realm having a `RealmUserDefault` entry. Probably the fix is to create an empty one for any organization that doesn't have one, either in the `zerver/data_import/` common library or in `import_realm.py`.

I'm having second thoughts about requiring every realm to have a `RealmUserDefault` entry.  We probably want to using this feature require Zulip Standard, and more generally we don't expect many users to actually set these.  And it might be kinda nice if the 98% of realms that don't use this feature can't be affected by potential bugs in the customizations we can do here?  Idunno, curious for your thoughts on that, and might be a good #backend thread topic.

sahil839 commented:
I agree that `UserProfile.property_types` or `RealmUserDefault.property_types`, have used `UserBaseSettings` for the `copy_default_settings` since there can be both type of source `UserProfile` and `RealmUserDefault`.

Still need to fix the import logic but.

Also you mentioned in the last comment that, you merged the `RealmUserDefault` table commit but I cannot see that.

timabbott commented:
Yeah, it looks like I merged only the commit before that one; did so just now: d7dfe80454e00d009ffee745ff5af1454e0b75a8.

The import commit fails tests if we merge it before the commit to create RealmUserDefault objects on realm creation, since the test database won't have one to `.get()` as of that time.

timabbott commented:
I also merged the `property_types` move commits in 2355e66850c7d52541584e74542a4af6d868b0e7.

sahil839 commented:
@timabbott Updated the PR. Made the following changes apart from solving merge conflicts -
- Handle the case when there is no `RealmUserDefault` table while importing by creating a new one.
- Changed the `do_set_realm_user_default_setting` to have RealmUserDefault object as a paramter instead of realm object to avoid querying database multiple times when updating multiple settings using API.

sahil839 commented:
We can hold `models: Move default-language and 24-hour-time to RealmUserDefault.` commit for now, as I am planning to have this commit at a later stage because I feel that would be easy.

timabbott commented:
Does it make sense to hold this until we've merged the `user_settings` event type migration, or will they rebase cleanly with respect to each other?  I feel pretty good about us being able to merge the `user_settings` migration this week if we push.

sahil839 commented:
Yes, would be better to do this after that migration, I think the conflicts will be resolved easily.

sahil839 commented:
Last push was only for rebasing, not ready for review yet.

sahil839 commented on zerver/lib/events.py:
```diff
@@ -322,6 +323,18 @@ def fetch_initial_state_data(
+            state["realm_user_settings_defaults"][property_name] = getattr(
+                realm_user_default, property_name
+            )
+
```
Here also we would require `available_notification_sounds` and `emojiset_choices` that we sent currently in `user_settings` object. But that should be same in both cases - for both user and realm settings. Do we want to duplicate them here or just remove it from `user_settings` object and include in the top-level object.

timabbott replied:
I'm not sure what the best approach is here; maybe start an thread in `#api design`?

sahil839 replied:
On thinking again, I think we would want to have two copies, as there can be case when `event_types` only includes `realm_user_settings_defaults` and also it would not be of use if none of `realm_user_settings_defaults` and `user_settings` is included in event types.

timabbott replied:
Sounds reasonable.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -10630,6 +10669,201 @@ paths:
+                            type: integer
+                            description: |
+                              The duration (in seconds) for which the server should wait to batch
+                              email notifications before sending them.
```
These are duplicated as `user_settings` has all these. I cannot figure out how to deduplicate this.

timabbott replied:
This is OK for now but clearly something we'll want to fix since it's a lot of duplicated content.  Can you open an issue for deduplicating the settings definitions in the API docs, and a second for deduplicating the `ignored_parameters_unsupported` text in the API docs (which is small but something we're planning to add to several more API endpoints in the future); you can mention @MSurfer20 on both issues since he's likely to find them fairly quick.

sahil839 commented on tools/linter_lib/custom_check.py:
```diff
@@ -889,7 +889,10 @@
+            "exclude": {
+                "templates/zerver/help/change-organization-url.md",
+                "templates/zerver/help/include/rest-endpoints.md",
+            },
```
This would result in someone using `realm` for the sidebar of the api docs, any other solution to use `/api/update-realm-user-settings-defaults` without getting error.

timabbott replied:
We probably should exclude `-realm-` with `exclude_pattern` as a cleaner solution for this.

sahil839 replied:
Done

sahil839 commented:
@timabbott ready for review. Left some comments above.
- Removed the commit for migrating existing settings for now.
- Update the variables and urls as discussed in chat.
- Also added documentation for the new endpoint.

If this is fine, then we can proceed with discussion for the UI.

sahil839 commented:
> @timabbott ready for review. Left some comments above.
> 
> * Removed the commit for migrating existing settings for now.
> * Update the variables and urls as discussed in chat.
> * Also added documentation for the new endpoint.
> 
> If this is fine, then we can proceed with discussion for the UI.

Previous push was just to rebase.

timabbott commented on templates/zerver/help/include/rest-endpoints.md:
```diff
@@ -77,6 +77,7 @@
 * [Get all custom profile fields](/api/get-custom-profile-fields)
 * [Reorder custom profile fields](/api/reorder-custom-profile-fields)
 * [Create a custom profile field](/api/create-custom-profile-field)
+* [Update default values of user settings](/api/update-realm-user-settings-defaults)
```
I'd call it "Change default values of user preferences".

timabbott commented on zerver/migrations/0341_create_realm_user_default_table.py:
```diff
@@ -0,0 +1,29 @@
+    # Zulip "system bots" and will have no use of this table.
+    realms = Realm.objects.exclude(string_id=settings.SYSTEM_BOT_REALM)
+    for realm in realms:
+        RealmUserDefault.objects.create(realm=realm)
```
It might be worth doing this as a bulk create so that we don't do a ton of database queries in a loop in Zulip Cloud.

sahil839 commented:
@timabbott updated the PR as per review.

sahil839 commented:
I have frontend work almost ready. Would be better if this is merged first after we decide the behavior for `enable_marketing_emails`.

sahil839 commented:
> @timabbott ready for review.
> I am not sure about how we should handle `enable_marketing_emails`, I have added commit to remove it from `notification_setting_types` and hence `RealmUserDefault` and have explained in commit message also. But after this change there is still question of what should be behavior when copying settings from other user profile, as of now we just set the value as selected in registration form and the source profile setting does not affect. There is some discussion at https://chat.zulip.org/#narrow/stream/2-general/topic/marketing.20emails.20enabled.20setting.20for.20new.20users.
> 
> Apart from the `enable_marketing_emails` commit, there is no change done after responding to the last review.

I added a commit for moving `emojiset_choices` to UserBaseSettings class. And I have two questions mainly - one about the `enable_marketing_emails` as quoted above and other about the `emojiset_choices` and `available_notification_sounds` fields in the `register response`, whether we should just inlcude it in the top-level object or have its two copies in `user_settings` and `realm_user_settings_defaults`.

timabbott commented:
OK I think this PR is very close to merging.  I made a few changes to comments etc. for readability and pushed back here.  I had been intending to finish an audit of the API documentation and merge this, but noticed a few questions that deserves more thought:
* Do we want RealmUserDefault to apply to bot users?  My theory is no, because
* I think we don't want to exclude `SYSTEM_BOT_REALM`, because there are actual users in that realm in Zulip Cloud, so it'd be an invariant violation to do so.  Does it work to just remove the exclusion?  (I think it won't matter in the long term since @mateuszmandera is planning to remove `SYSTEM_BOT_REALM` as a concept soon anyway).
* I added various comments on the `enable_marketing_emails` code path.  I wanted to ask if we'd prefer an alternative implementation as follows: (1) we include it in RealmUserDefault as normal and thus remove all the custom code for the setting, (2) we change the `create_user` code path to set `user_profile.enable_marketing_emails` based on the passed parameter after the copy step, (3) we make sure we pass through `None` rather than True/False as the default for the case that one didn't specify anything in the registration form, and have `create_user` do nothing with a `None` value passed in.  The net effect would be that the `enable_marketing_emails` setting on `RealmUserDefault` would just need to be documented as having no effect unless it's an SSO situation with `TERMS_OF_SERVICE=False`, where one wouldn't normally be presented with a registration form at all (though the setting probably doesn't do anything in that case anyway, so ... it really doesn't matter).  If this is complicated to switch or think through switching, how this works is pretty low priority one way or the other, so it'd be better to just merge what we have and move on (especially since I've already read this PR carefully); we can always revisit that detail later.  But my overall preference is more to avoid having this setting be a weird exception with custom code in like 10 places; it'd be better to just have it be an exception in like one case if we can.  (E.g. have the view function return an error if you try to toggle it).

sahil839 commented:
Updated the PR.
- The settings from `RealmUserDefault` are now set only for non-bot users.
- We create the `RealmUserDefault` table for `SYSTEM_BOT_REALM` as well.
- Added `enable_marketing_emails` to `notification_setting_types`. Had one doubt in this whether we should include this setting in `realm_user_settings_defaults` object of `register` response.
- Updated the code to set `enable_marketing_emails` after copying the user settings and setting the RealmUserDefault values.
- Added commit to set default object as None. I have not been able to test the None case yet. I tried Github and Gitlab SSO (with `TERMS_OF_SERVICE` as False) and for both registration form needs to be filled with name and `enable_marketing_emails` fields. I also tried google SSO but there was some error when trying to sign-in.

timabbott commented:
This looks great; I think it's worth stalling another day for you to clean up the `property_types` usage (I think there's some redundancy happening where we process notification settings twice right now).  Beyond that I have a few small changes to documentation in mind.

I think duplicating the `emojiset_choices` type objects is probably fine in the short term; in the long term we likely want to move them to their own top-level object anyway, and so whatever is simplest to implement is probably best for now.

timabbott commented on zerver/migrations/0344_create_realm_user_default_table.py:
```diff
@@ -0,0 +1,31 @@
+    Realm = apps.get_model("zerver", "Realm")
+    RealmUserDefault = apps.get_model("zerver", "RealmUserDefault")
+    # We exclude the "SYSTEM_BOT_REALM" here because this realm only contains
+    # Zulip "system bots" and will have no use of this table.
```
This comment should be deleted, since it's no longer accurate.

sahil839 replied:
Done

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -3884,6 +3884,45 @@ paths:
+                              description: |
+                                Event sent to all users in a Zulip organization when the
+                                default settings for new users of the organization
+                                (realm) has changed.
```
"have changed"

sahil839 replied:
Done

timabbott commented on zerver/views/realm.py:
```diff
@@ -275,3 +278,109 @@ def realm_reactivation(request: HttpRequest, confirmation_key: str) -> HttpRespo
+    email_notifications_batching_period_seconds: Optional[int] = REQ(
+        json_validator=check_int, default=None
+    ),
+) -> HttpResponse:
```
Medium term, we will want to figure out how to avoid this huge set of variables being duplicated with the `user_settings.py` endpoint.  I don't have a specific plan at this time.

timabbott replied:
One idea is we can have a `user_settings_request_dict = parse_user_settings_request(request)` function which has the `@has_request_varibles` decorator and returns effectively its `locals()` filtered by what is present in `property_types`, and call that function from both here and the `user_settings.py` endpoint.

timabbott commented on zerver/views/realm.py:
```diff
@@ -275,3 +278,109 @@ def realm_reactivation(request: HttpRequest, confirmation_key: str) -> HttpRespo
+            _("Invalid email batching period: {} seconds").format(
+                email_notifications_batching_period_seconds
+            )
+        )
```
It might be useful to refactor these two validation checks into a function (or even a `validator`, so that we avoid duplicating this code with `user_settings.py`.

sahil839 replied:
Have done this but have left a comment regarding location of this function. I am not sure what you meant by validator here. I guessed it to be something which can be passed to REQ as json_validator, but a comment on top says that we cannot do the check for `default_language` in REQ.

timabbott replied:
Hmm, yeah.  I bet we could figure out a way to remove that constraint (E.g. having the allowed languages be the empty set if we haven't provisioned, rather than having it crash), but I don't see a lot of urgency to do so.

timabbott commented on zerver/views/realm.py:
```diff
@@ -275,3 +278,109 @@ def realm_reactivation(request: HttpRequest, confirmation_key: str) -> HttpRespo
+        k: v
+        for k, v in list(locals().items())
+        if (k in user_profile.property_types or k in realm_user_default.notification_setting_types)
+    }
```
(We can delete this first loop, I think).

sahil839 replied:
I am not sure what you meant by first lopp here, I have removd the `k in realm_user_default.notification_setting_types` condition.

timabbott replied:
Yeah, that's what I meant.

sahil839 commented on zerver/lib/validator.py:
```diff
@@ -582,3 +582,36 @@ def check_string_or_int(var_name: str, val: object) -> Union[str, int]:
+    default_language: Optional[str],
+) -> None:
+    from zerver.lib.actions import get_available_notification_sounds
+    from zerver.lib.i18n import get_available_language_codes
```
I have imported it here because tests were failing due to cyclic imports. Is there any other file where we can keep this function such that it will be used both in `views/realm.py` and `views/user_settings.py` if we want import statements at top.

timabbott replied:
I think this could just live in `zerver/views/user_setttings.py`, and the `realm` code could import that.

timabbott replied:
(Leaving this as a follow-up, since I just finished reviewing this whole giant PR and this is an easy follow-up cleanup).

sahil839 commented:
@timabbott updated the PR.
- Added commit for extracting the checks as a function (left an inline comment as well above).
- Added `emojiset_choices` and `available_notification_sounds` fields to `realm_user_settings_defaults` objects.
- Done all other changes mentioned above and also left a comment for one of those.

Apart from above, I wanted to confirm whether we want to include `enable_marketing_emails` in the `realm_user_settings_defaults` and also whether to allow changing it using `do_set_realm_user_default_setting`. We do not allow to change `enable_marketing_emails` using API endpoint though.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -1009,6 +1009,42 @@ def do_set_realm_signup_notifications_stream(
+    event_time = timezone_now()
+    RealmAuditLog.objects.create(
+        realm=realm,
+        event_type=RealmAuditLog.REALM_DEFAULT_PROPERTY_CHANGED,
```
I renamed this to `REALM_DEFAULT_USER_SETTINGS_CHANGED`

timabbott commented on zerver/views/realm.py:
```diff
@@ -275,3 +278,94 @@ def realm_reactivation(request: HttpRequest, confirmation_key: str) -> HttpRespo
+
+    realm_user_default = RealmUserDefault.objects.get(realm=user_profile.realm)
+    request_settings = {
+        k: v for k, v in list(locals().items()) if (k in user_profile.property_types)
```
Changed `user_profile.property_types` to `RealmUserDefault.property_types`, since that's more exactly the set that we want to look at here.

(One can imagine a world where those diverge for some reason).

timabbott commented:
Merged, thanks @sahil839!  I noted a few small changes I made while merging above; mainly to documentation details (E.g. I added a comment about `enable_marketing_emails` in the new endpoint) and also squashing the API documentation commit into the API feature commit, since there didn't seem to be any benefit to those being separate.  

Huge thanks for doing this @sahil839!  I think there's a handful of small follow-ups that we've discussed here, but otherwise I think that all that remains for this feature is the frontend (and /help/ docs).
